File: 1367-linked-list-in-binary-tree/1367-linked-list-in-binary-tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSubPath(ListNode* head, TreeNode* root) {
        if (!root) return false;
        // Check the current node and all its descendants
        bool result = dfs(root, head);
        // Check all paths from the left and right children of the root
        result |= isSubPath(head, root->left);
        result |= isSubPath(head, root->right);
        return result;
    }

private:
    bool dfs(TreeNode* node, ListNode* head) {
        if (!head) return true;
        if (!node) return false;
        bool result = false;
        if (node->val == head->val) {
            // Continue searching in both left and right subtrees
            result |= dfs(node->left, head->next);
            result |= dfs(node->right, head->next);
        }
        return result;
    }
};

File: 1367-linked-list-in-binary-tree/1367-linked-list-in-binary-tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public boolean isSubPath(ListNode head, TreeNode root) {
        if (root == null) return false;
        // Check the current node and all its descendants
        boolean result = dfs(root, head);
        // Check all paths from the left and right children of the root
        result = result || isSubPath(head, root.left);
        result = result || isSubPath(head, root.right);
        return result;
    }

    private boolean dfs(TreeNode node, ListNode head) {
        if (head == null) return true;
        if (node == null) return false;
        boolean result = false;
        if (node.val == head.val) {
            // Continue searching in both left and right subtrees
            result = dfs(node.left, head.next);
            result = result || dfs(node.right, head.next);
        }
        return result;
    }
}